Physics server plug

AndreaCatania opened the pull request:
I've implemented the possibility to plug the physics server.

To register new physics server you have to register it in this way: 

> ClassDB::register_class<PhysicsServerSW>();
> PhysicsServerManager::register_class<PhysicsServerSW>();

Optionally you can set new default physics server by using: 
> PhysicsServerManager::set_default_class(int p_priority = 0);

New server can be put into the "modules" directory.

After the registration of server you can select the one to use in the project settings under physics 3D.

---
@Reduz As you can see the default "Godot physics" is registered by register_server_types and all other servers can be set into modules and registered by its own registering function.
For this reason is necessary start the physics server after the registration of both servers and modules, so all registered physics server are taken
So I created these new functions: 

**FILE:** core/os/os.h
> virtual void initialize_physics() = 0;

**FILE:** servers/register_server_types.h
> void register_server_singletons();

Here the code where is handled the physics start: https://github.com/AndreaCatania/godot/blob/36e6024dcbc036c43a722cac9949ff3e1f3edde9/main/main.cpp#L1075-L1084

I've checked and my modification doesn't make any problem on my machine, but I'm not 100% sure if the registration of server's singletons and the physics server after the module registration may cause some problems, So @reduz please let me know if it's wrong or it's fine.

akien-mga commented on servers/physics_server.cpp:
```diff
@@ -732,3 +734,43 @@ PhysicsServer::~PhysicsServer() {
 	singleton = NULL;
 }
+
+Vector<String> PhysicsServerManager::physics_servers_classes;
```
Might be the C++ noob in me talking, but why should it be redeclared here if it's already in the header? Also, the initializations should typically go in a constructor, no?

AndreaCatania replied:
This is a static variable, and this is the way how to declare it

akien-mga commented on servers/physics_server.h:
```diff
@@ -658,6 +658,43 @@ class PhysicsServer : public Object {
+	template <class T>
+	static void register_class() {
+
+		ERR_FAIL_COND(-1 != find_class_id(T::get_class_static()));
```
It's a matter of preference, but we use `(stuff != -1)` everywhere else in Godot, so I'd prefer if you used it here too.

AndreaCatania replied:
Ok Thanks for advice

reduz commented on servers/physics_server.h:
```diff
@@ -658,6 +658,26 @@ class PhysicsServer : public Object {
 };
 
+class PhysicsServerManager {
+	static Vector<String> physics_servers_classes;
```
Just register function pointers with a name, using ClassDB names will restrict it from working using GDNative eventually if we want to provide external physics engines (such as PhysX or Havok)

AndreaCatania replied:
What do you advise me to use?

reduz commented on servers/register_server_types.cpp:
```diff
@@ -144,6 +145,9 @@ void register_server_types() {
 	ScriptDebuggerRemote::resource_usage_func = _debugger_get_resource_usage;
+
+	PhysicsServerManager::register_class(PhysicsServerSW::get_class_static());
+	PhysicsServerManager::set_default_class(PhysicsServerSW::get_class_static());
```
Again, just use a function pointer that creates a PhysicsServer, this will allow it to work in GDNative too, which has no access to ClassDB.

reduz commented on core/project_settings.cpp:
```diff
@@ -1044,6 +1045,10 @@ ProjectSettings::ProjectSettings() {
 	using_datapack = false;
+
+	// Physics settings
+	Compression::zstd_level = GLOBAL_DEF(PhysicsServerManager::setting_property_name, "DEFAULT");
```
this doesnt make any sense

AndreaCatania replied:
Do you mean that the assignment "Compression::zstd_level = GLOBAL_DEF" is useless or the definition of "DEFAULT"?

My idea was to allow the developer to set the default physics server directly in the code. So eventually is not required by the "Godot" user that change the settings in order to use the default physics.

akien-mga replied:
What's the link with `Compression::zstd_level` though?

akien-mga commented:
After addressing Juan's comments, please squash the commits together, one should be enough for this diff.

reduz commented on core/os/os.h:
```diff
@@ -116,6 +118,8 @@ class OS {
 	virtual void initialize_logger();
 	virtual void initialize_core() = 0;
+	PhysicsServer *new_physics_server();
+	virtual void initialize_physics() = 0;
```
Actually this is a really good idea, but I would just make this function non abstract and put the initialization code here, since it seems unlikely it may change on different OSs.

akien-mga replied:
That part still has to be done I think?

AndreaCatania commented:
@reduz I've changed the code, now it doesn't use ClassDB as you suggested. The only think that I don't know how to do is that 

```
PhysicsServer *new_physics_server();
virtual void initialize_physics() = 0;
```

> Actually this is a really good idea, but I would just make this function non abstract and put the initialization code here, since it seems unlikely it may change on different OSs.

Each OS has its implementation that initialize physics 2d. If you tell me which I should use I can do it. Another thing is that each os has its physics_server object, I'll move it inside the OS class.

AndreaCatania committed:
Implemented physics plug

Moved init_physics

Implemented physics 2D plug

Fix clang

Fix clang

Fix static check

Fix clang

Fix static check

Moved physics server initialization

Moved physics server settings initialization

akien-mga commented:
It fails building on OSX:
```
platform/osx/godot_main_osx.mm:83:9: error: variable type 'OS_OSX' is an abstract class
        OS_OSX os;
               ^
core/os/os.h:122:15: note: unimplemented pure virtual method 'initialize_physics' in 'OS_OSX'
        virtual void initialize_physics() = 0;
                     ^
```

reduz commented on platform/android/os_android.cpp:
```diff
@@ -133,6 +133,14 @@ void OS_Android::set_opengl_extensions(const char *p_gl_extensions) {
 	gl_extensions = p_gl_extensions;
 }
 
+void OS_Android::initialize_physics() {
```
As this code is the same in all OSs, i would just leave it at OS.

AndreaCatania commented:
@reduz @akien-mga I fixed the requested things, let me know

akien-mga commented on platform/windows/crash_handler_win.cpp:
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 #include "main/main.h"
 #include "os_windows.h"
+#include "project_settings.h"
```
Why?

akien-mga replied:
Nevermind, I see why now :)

AndreaCatania commented:
Finally I fixed all compilation problems.

So now we have both 2D and 3D physics pluggable

reduz commented on core/os/os.cpp:
```diff
@@ -73,6 +75,35 @@ void OS::initialize_logger() {
 	_set_logger(memnew(StdLogger));
 }
 
+void OS::initialize_physics() {
```
I thought we discussed moving this code to main.cpp , similar to how AudioServer works, to a void a dependency from core to servers

AndreaCatania commented:
Yes but I resolved by creating physical plug either for 2D. So is required
only the inclusion of physical server.

The reason why I leaved it there is because the physics server variable is
inside the OS and I think that initialize and destruct inside the OS class
is the right way.

Or do you want to move the physics servers variables inside the OS?

On Nov 3, 2017 8:52 PM, "Juan Linietsky" <notifications@github.com> wrote:

> *@reduz* commented on this pull request.
> ------------------------------
>
> In core/os/os.cpp
> <https://github.com/godotengine/godot/pull/12262#discussion_r148878806>:
>
> > @@ -73,6 +75,35 @@ void OS::initialize_logger() {
>  	_set_logger(memnew(StdLogger));
>  }
>
> +void OS::initialize_physics() {
>
> I thought we discussed moving this code to main.cpp , similar to how
> AudioServer works, to a void a dependency from core to servers
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/12262#pullrequestreview-74195664>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH9MR1om8KWnJrWKxeSkCRc3F6rGrS9zks5sy27agaJpZM4QA0tF>
> .
>

reduz commented:
@AndreaCatania Well, as I mentioned. We can't have core depend on servers. Godot is very strict on dependency between libraries. This is why I suggested moving the code to main.cpp and initialize/uninitialize like AudioServer. This would make OS not relate to PhysicsServer at all.

AndreaCatania commented:
@reduz I did the modification, now should be perfect, let me know

reduz commented on core/project_settings.cpp:
```diff
@@ -1044,6 +1046,14 @@ ProjectSettings::ProjectSettings() {
 	custom_prop_info["compression/formats/gzip/compression_level"] = PropertyInfo(Variant::INT, "compression/formats/gzip/compression_level", PROPERTY_HINT_RANGE, "-1,9,1");
 
 	using_datapack = false;
+
```
This also is core using servers, but should go fine in main.cpp too. you can call it before setting up the servers.

reduz reviewed (changes requested):
Seems mostly fine, would just move that ProjectSettings line to main.cpp too

AndreaCatania commented:
@reduz I changed it

reduz commented:
Finally merged, hooray! Sorry that I'm extremely picky when it comes to core, as this is the most delicate part of the engine.

AndreaCatania commented:
@reduz Yes it's fine for me :)
